src/libsyntax/parse/parser.rs

```diff
@@ -3992,7 +3992,9 @@ impl<'a> Parser<'a> {
         // Starts like a simple path, but not a union item.
         } else if self.token.is_path_start() &&
                   !self.token.is_qpath_start() &&
-                  !self.is_union_item() {
+                  !self.is_union_item() &&
+                  !(self.token.is_keyword(keywords::Crate) &&
+                    self.look_ahead(1, |t| t != &token::ModSep)) {
```

Comment:
Nit: I feel like some comments would be nice here too. In fact, I personally am a bit confused about what's going on here. =) Lots of nested nots! 

```
// Parse any path `foo::bar`. But be careful of various corner cases involving contextual keywords:
//
// - `union Foo { .. }`, a union item. This gets parsed below. Note that `union::foo` is a path.
// - Note that `crate::foo` *is* a path, but `crate` followed by any other token is not. (??)
```

I'm not entirely clear I guess on what is happening. =)